primitives/s2s/src/call_params.rs

```diff
@@ -0,0 +1,193 @@
+// This file is part of Darwinia.
+//
+// Copyright (C) 2018-2021 Darwinia Network
+// SPDX-License-Identifier: GPL-3.0
+//
+// Darwinia is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Darwinia is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Darwinia. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+// --- paritytech ---
+use bp_message_dispatch::CallOrigin;
+use bp_runtime::messages::DispatchFeePayment;
+use sp_core::{H160, U256};
+use sp_runtime::{AccountId32, MultiSignature, MultiSigner};
+use sp_std::{vec, vec::Vec};
+// --- darwinia-network ---
+use codec::{Decode, Encode};
+use dp_asset::token::TokenMetadata;
+
+/// The parameters box for the pallet runtime call.
+#[derive(Encode, Decode, Debug, PartialEq, Eq, Clone)]
+pub enum CallParams {
+	#[codec(index = 0)]
+	S2sIssuingPalletRegisterFromRemote(TokenMetadata),
+	#[codec(index = 1)]
+	S2sIssuingPalletIssueFromRemote(H160, U256, H160),
+	#[codec(index = 2)]
+	S2sBackingPalletUnlockFromRemote(H160, U256, Vec<u8>),
+}
+/// Creating a concrete message payload which would be relay to target chain.
+pub trait CreatePayload<SourceChainAccountId, TargetChainAccountPublic, TargetChainSignature>
+where
+	SourceChainAccountId: Encode,
+	Self: Sized,
```

Comment:
Following is payload's type, I'm sure whether we can add more limitation to Payload associate type:
Aka. geting types <SourceChainAccountId, TargetChainAccountPublic, TargetChainSignature> from MessagePayload

```
/// Message payload type used by dispatch module.
#[derive(RuntimeDebug, Encode, Decode, Clone, PartialEq, Eq)]
pub struct MessagePayload<
	SourceChainAccountId,
	TargetChainAccountPublic,
	TargetChainSignature,
	Call,
> {
	/// Runtime specification version. We only dispatch messages that have the same
	/// runtime version. Otherwise we risk to misinterpret encoded calls.
	pub spec_version: SpecVersion,
	/// Weight of the call, declared by the message sender. If it is less than actual
	/// static weight, the call is not dispatched.
	pub weight: Weight,
	/// Call origin to be used during dispatch.
	pub origin: CallOrigin<SourceChainAccountId, TargetChainAccountPublic, TargetChainSignature>,
	/// Where the fee for dispatching message is paid?
	pub dispatch_fee_payment: DispatchFeePayment,
	/// The call itself.
	pub call: Call,
}
```

Reply:
Self: Sized this line restricts the type CreatePayload itself. Not its inner type Payload. Is it necessary?

Reply:
It's not necessary indeed, include the `SourceChainAccountId: Encode`.